File: go/problem/three_sum.go
package problem

import "sort"

func threeSum(nums []int) [][]int {
	var result [][]int

	sort.Ints(nums)
	for i := 0; i < len(nums)-2; i++ {
		if nums[i] >= 1 {
			break
		}

		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		l, r := i+1, len(nums)-1
		for l < r {
			sum := nums[i] + nums[l] + nums[r]
			switch {
			case sum == 0:
				result = append(result, []int{nums[i], nums[l], nums[r]})
				for l++; l < r && nums[l] == nums[l-1]; l++ {
				}
				for r--; l < r && nums[r] == nums[r+1]; r-- {
				}

			case sum > 0:
				r--
			default:
				l++
			}
		}
	}

	return result
}

File: rust/src/problems/three_sum.rs
use std::cmp::Ordering;

pub struct Solution;
impl Solution {
    pub fn three_sum(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut result = Vec::new();
        let mut sorted_nums = nums.clone();
        sorted_nums.sort();

        for i in 0..sorted_nums.len() {
            if sorted_nums[i] > 0 {
                break;
            }
            if i > 0 && sorted_nums[i] == sorted_nums[i - 1] {
                continue;
            }
            let mut left = i + 1;
            let mut right = sorted_nums.len() - 1;

            while left < right {
                let sum = sorted_nums[i] + sorted_nums[left] + sorted_nums[right];
                match sum.cmp(&0) {
                    Ordering::Equal => {
                        result.push(vec![sorted_nums[i], sorted_nums[left], sorted_nums[right]]);
                        while left < right && sorted_nums[left] == sorted_nums[left + 1] {
                            left += 1;
                        }
                        while left < right && sorted_nums[right] == sorted_nums[right - 1] {
                            right -= 1;
                        }
                        left += 1;
                        right -= 1;
                    }
                    Ordering::Less => left += 1,
                    Ordering::Greater => right -= 1,
                }
            }
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_max_profit() {
        let nums = vec![-1, 0, 1, 2, -1, -4];
        let expected = vec![vec![-1, -1, 2], vec![-1, 0, 1]];
        assert_eq!(Solution::three_sum(nums), expected);

        let nums = vec![1, 2, 3];
        let expected: Vec<Vec<i32>> = Vec::new();
        assert_eq!(Solution::three_sum(nums), expected);

        let nums: Vec<i32> = Vec::new();
        let expected: Vec<Vec<i32>> = Vec::new();
        assert_eq!(Solution::three_sum(nums), expected);

        let nums = vec![-1, 0, 1, 2, -1, -4, 1, -1, -1, 0];
        let expected = vec![vec![-1, -1, 2], vec![-1, 0, 1]];
        assert_eq!(Solution::three_sum(nums), expected);

        let nums = vec![-1, 0, 1, 2, -1, -4];
        let expected = vec![vec![-1, -1, 2], vec![-1, 0, 1]];
        assert_eq!(Solution::three_sum(nums), expected);
    }
}
